hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java

```diff
@@ -0,0 +1,422 @@
+            boolean fetched = poll(this::emit);
+            if (fetched) {
+                idleCount = 0;
+            } else {
+                IDLER.idle(++idleCount);
+            }
+        }
+        return true;
+    }
+
+    private boolean poll(Consumer<T> consumer) {
+        Map<Integer, ICompletableFuture<ReadResultSet<T>>> futureMap = new HashMap<>();
+        for (Map.Entry<Integer, Long> entry : offsetMap.entrySet()) {
+            int partitionId = entry.getKey();
+            long offset = entry.getValue();
+            ICompletableFuture<ReadResultSet<T>> future = eventJournalReader.readFromEventJournal(offset,
+                    MIN_FETCH_SIZE, MAX_FETCH_SIZE, partitionId, predicate, projection);
+            futureMap.put(partitionId, future);
+        }
+        boolean fetched = false;
+        for (Map.Entry<Integer, ICompletableFuture<ReadResultSet<T>>> entry : futureMap.entrySet()) {
+            Integer partitionId = entry.getKey();
+            ICompletableFuture<ReadResultSet<T>> future = entry.getValue();
+            try {
+                ReadResultSet<T> resultSet = future.get();
+                for (T event : resultSet) {
+                    consumer.accept(event);
+                    fetched = true;
+                }
+                offsetMap.compute(partitionId, (k, v) -> v + resultSet.readCount());
+            } catch (ExecutionException e) {
+                if (e.getCause() instanceof StaleSequenceException) {
+                    long headSeq = ((StaleSequenceException) e.getCause()).getHeadSeq();
+                    long oldOffset = offsetMap.put(partitionId, headSeq);
+                    System.out.println("fatal stale sequence, requested: " + oldOffset + ", currentHead: " + headSeq);
+                    continue;
+                }
+                throw ExceptionUtil.rethrow(e);
+            } catch (InterruptedException e) {
+                throw ExceptionUtil.rethrow(e);
+            }
+        }
+        return fetched;
+    }
+
+    @Override
+    public boolean isCooperative() {
+        return false;
+    }
+
+    private static <E, T> Projection<E, T> createProjection(Function<E, T> projectionF) {
+        if (projectionF == null) {
+            return null;
+        }
+        return new Projection<E, T>() {
+            @Override public T transform(E input) {
+                return projectionF.apply(input);
+            }
+        };
+    }
+
+    private static <E, T> List<Processor> getProcessors(int count, List<Integer> ownedPartitions,
+                                                        EventJournalReader<E> eventJournalReader,
+                                                        Predicate<E> predicate,
+                                                        Function<E, T> projection,
+                                                        boolean newestSequence) {
+        Map<Integer, List<Integer>> processorToPartitions = range(0, ownedPartitions.size())
```

Comment:
maybe this can be extracted to a utility function? same logic duplicated in `ReadWithPartitionIteratorP`